### PR TITLE
General rate limiting improvements

### LINKED ISSUE #161: Remove unnecesary wait after stop signal is received
Currently, we wait for one more "cycle" after the stop signal is received. This is not needed, it prolongs tests and is in general annoying. We can remove it and (presumably) shorten some E2E tests.

### LINKED ISSUE #308: Stagger API calls when possible
Airseeker calls APIs, Airnode gateways and blockchain providers to read things. It does these in bursts, which creates instantaneous load. If these calls are staggered, it would help. However, these two future features need to be considered: https://github.com/api3dao/airseeker/issues/256 https://github.com/api3dao/airseeker/issues/216

### PR DESCRIPTION
**Pull request by aquarat**
This PR:
- Adds Bottleneck to provider requests and makes the limiters configurable globally (but each provider has its own limiter)
  - Provider rate limiting is also configurable per RPC URL 
- Adds Bottleneck to direct gateway requests (as above)
  - Direct gateway limiters are configured globally (with one limiter per OIS) but config can also be overridden per ois title
- Adds Bottleneck to signed data gateway requests (as above)
  - One limiter per airnode address, configured globally but also configurable per airnode address
- Bumps Node version from 14 to 16 in CI and package.json
- Hastily terminates if CTRL+C sent twice
- Configures ethers Provider to have a socket timeout that is less than the go-utils timeout minus a small headroom window (ensure the socket is closed gracefully/not left dangling)

Closes #308 
Closes #161 

I think @bdrhn9 will have to test this as testing this without the Nodary config file and providers could be difficult.

Example airseeker config excerpt:
```json
{
  "airseekerWalletMnemonic": "${AIRSEEKER_WALLET_MNEMONIC}",
  "rateLimiting": {
    "maxGatewayConcurrency": 5,
    "minGatewayTime": 200,
    "maxProviderConcurrency": 5,
    "minProviderTime": 200,
    "minDirectGatewayTime": 200,
    "maxDirectGatewayConcurrency": 5,
    "overrides": {
      "signedDataGateways": {
        "0xA30CA71Ba54E83127214D3271aEA8F5D6bD4Dace": {
          "minTime": 200,
          "maxConcurrent": 5
        }
      },
      "directGateways": {
        "Currency Converter API": {
          "minTime": 200,
          "maxConcurrent": 5
        }
      }
    }
  },
  "log": {
    "format": "plain",
    "level": "INFO"
  },
  "chains": {
    "3": {
      "contracts": {
        "Api3ServerV1": "0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512"
      },
      "providers": {
        "infuraRopsten": {
          "url": "${CP_INFURA_ROPSTEN_URL}",
          "rateLimiter": {
            "minTime": 200,
            "maxConcurrent": 5
          }
        }
      },
      "options": {
        "fulfillmentGasLimit": 500000,
        "gasPriceOracle": [
          {
            "gasPriceStrategy": "latestBlockPercentileGasPrice",
            "percentile": 60,
            "minTransactionCount": 20,
            "pastToCompareInBlocks": 20,
            "maxDeviationMultiplier": 2
          },
          {
            "gasPriceStrategy": "providerRecommendedGasPrice",
            "recommendedGasPriceMultiplier": 1.2
          },
          {
            "gasPriceStrategy": "constantGasPrice",
            "gasPrice": {
              "value": 10,
              "unit": "gwei"
            }
          }
        ]
      }
    }
}
```

### CODE REVIEW

**Comment by bdrhn9** [comment]
> I think @bdrhn9 will have to test this as testing this without the Nodary config file and providers could be difficult.

Is it something bad to test this with updated Nodary config file only on testnets?

**Comment by aquarat** [comment]
@bdrhn9 
> Is it something bad to test this with updated Nodary config file only on testnets?

That sounds perfect. It just needs to be tested in a production-like environment (which is what testnets are for).
I'm just finishing up the validation and then it'll be ready for testing.

**Comment by aquarat** [comment]
@bdrhn9 The current defaults for everything are 200ms `minTime` and 5 `maxConcurrent`. 
- `maxConcurrent` can probably be increased for gateways to whatever the AWS concurrency is on the remote account minus some headroom (20?)
- `minTime` can probably be set to 1 or 0, in which case only the concurrency will matter.
- Defaults are probably too conservative

File: src/constants.ts [hunk 1]
```diff
@@ -18,3 +18,30 @@ export const NO_FETCH_EXIT_CODE = 2;
...
 // into 224 bits, its multiplication by 10^8 is guaranteed not to overflow.
 export const HUNDRED_PERCENT = 1e8;
+
+// The difference between the socket timeout (used in axios utils) and the go-utils total timeout.
+// This prevents dangling sockets which ultimately cause file descriptor exhaustion.
+export const TOTAL_TIMEOUT_HEADROOM = 500;
+
+// The difference between the socket timeout (used in ethers) and the go-utils total timeout.
+// This prevents dangling sockets which ultimately cause file descriptor exhaustion.
+export const PROVIDER_TIMEOUT_HEADROOM_MS = 500;
```
Comment (acenolaza) [inline src/constants.ts]: I'm just curious to know why this const has the `_MS` suffix and other time related ones don't 😄

Comment (acenolaza) [reply src/constants.ts]: Something similar I think is happening with `_DEFAULT` suffix below

File: src/state.ts [hunk 2]
```diff
@@ -33,17 +43,100 @@ export const initializeState = (config: Config) => {
...
+    return undefined;
+  }
+
+  return signedGatewayOverrides[airnodeAddress];
+};
+
+export const buildGatewayLimiters = (gateways?: Record<string, Gateway[]>, config?: Config) =>
+  gateways
+    ? Object.fromEntries(
+        Object.entries(gateways).map(([key, gateways]) => [
```
Comment (acenolaza) [inline src/state.ts]: Nit:
```suggestion
        Object.entries(gateways).map(([airnodeAddress, gateways]) => [
```

File: src/make-request.ts [hunk 3]
```diff
@@ -29,34 +30,42 @@ export function signWithTemplateId(templateId: string, timestamp: string, data:
...
         },
-        data: { encodedParameters: parameters },
-        timeout: GATEWAY_TIMEOUT_MS,
-      });
+        { totalTimeoutMs: GATEWAY_TIMEOUT_MS - TOTAL_TIMEOUT_HEADROOM }
+      );
 
-      return data;
-    });
+    const goRes = await (queue?.schedule ? queue.schedule(goResFn) : goResFn());
```
Comment (acenolaza) [inline src/make-request.ts]: Nit: the object that can be `undefined` is `queue`, right?
```suggestion
    const goRes = await (queue ? queue.schedule(goResFn) : goResFn());
```

File: src/make-request.ts [hunk 4]
```diff
@@ -114,11 +124,17 @@ export const makeApiRequest = async (template: Id<Template>): Promise<SignedData
...
     ...endpoint,
   };
-  const [_, apiCallResponse] = await node.api.callApi({
-    type: 'http-gateway',
-    config: { ois, apiCredentials },
-    aggregatedApiCall,
-  });
+
+  const limiter =
+    apiLimiters && apiLimiters[template.id] ? apiLimiters[template.id] : { schedule: (arg: any) => arg() };
```
Comment (acenolaza) [inline src/make-request.ts]: Nit: 

```suggestion
const limiter = apiLimiters?.[template.id] ?? { schedule: (arg: any) => arg() };
```

Comment (acenolaza) [reply src/make-request.ts]: Now that I think about it, should we create a new Bottleneck object if apiLimiters using the default const values is `undefined` 🤔

File: src/providers.ts [hunk 5]
```diff
@@ -1,9 +1,81 @@
...
+    // The parent must time out after the child scope to avoid leaking file descriptors.
+    //
+    // 3rd go-utils: PROVIDER_TIMEOUT_MS = 5_000 ms
+    // 2nd Bottleneck: PROVIDER_TIMEOUT_MS-(PROVIDER_TIMEOUT_HEADROOM_MS/2) = 5_000 - 500/2 = 5_000 - 250 = 4_750 ms
+    // 1st ethers socket: PROVIDER_TIMEOUT_MS-PROVIDER_TIMEOUT_HEADROOM_MS = 5_000 - 500 = 4_500 ms
+    return poll(() =>
+      this.limiter.schedule({ expiration: PROVIDER_TIMEOUT_MS - PROVIDER_TIMEOUT_HEADROOM_MS / 2 }, () =>
+        super.perform(method, params).then((result) => result, Promise.reject)
+      )
+    );
```
Comment (acenolaza) [inline src/providers.ts]: This is really cool and mind-blowing 🤯  but don't you need to pass the [PollOptions](https://github.com/ethers-io/ethers.js/blob/v5.7/packages/web/src.ts/index.ts#L407) object as a second argument to the function in order to accomplish what is stated in the comment on line 36?

Also, there seems to be a v6 release of ethers.js

**Review by acenolaza — APPROVED** [review]
LGTM 🔥 Good job with the custom ethers.js provider 👏🏻 I added a question around that topic and other really minor comments that can be ignored.

**Comment by bdrhn9** [comment]
> @bdrhn9 The current defaults for everything are 200ms `minTime` and 5 `maxConcurrent`.
> 
>     * `maxConcurrent` can probably be increased for gateways to whatever the AWS concurrency is on the remote account minus some headroom (20?)
> 
>     * `minTime` can probably be set to 1 or 0, in which case only the concurrency will matter.
> 
>     * Defaults are probably too conservative

I couldn't make it work even using half of 107 beacons on only testnets. Got timeout errors consistently. Tired to tune parameters but couldn't be successful. I can share my config, you'll test it with dummy mnemonic. Fyi, I didn't filtered empty sponsor wallets.

File: src/constants.ts (hunk 1 above)
Comment (aquarat) [reply src/constants.ts]: Thanks, I believe I've corrected them now :)

File: src/providers.ts (hunk 5 above)
Comment (aquarat) [reply src/providers.ts]: ah yes, you're right

File: src/make-request.ts (hunk 4 above)
Comment (aquarat) [reply src/make-request.ts]: Yes, good idea 👌

File: src/providers.ts (hunk 5 above)
Comment (aquarat) [reply src/providers.ts]: The `Poll`ing may actually not be necessary, removed pending testing :)

**Comment by aquarat** [comment]
This should be ready for testing again @bdrhn9 :) 🙏 

You'll want to use this to start with for rate limiting config:
```json
  "rateLimiting": {
    "maxGatewayConcurrency": 5,
    "minGatewayTime": 10,
    "maxProviderConcurrency": 5,
    "minProviderTime": 10,
    "maxDirectGatewayConcurrency": 5,
    "minDirectGatewayTime": 10
  },
```
^ I think the max concurrencies are the most important entries. The minimum times can probably be zero even.

**Comment by aquarat** [comment]
> LGTM 🔥 Good job with the custom ethers.js provider 👏🏻 I added a question around that topic and other really minor comments that can be ignored.

I forgot to say thanks for the feedback, turns out the Poll function was unnecessary :)

**Comment by aquarat** [comment]
Currently testing this with the beacon sets Mertcan created and it seems to be working well. I can see the issue Michal K was talking about with the gateways not responding under load.

**Comment by acenolaza** [comment]
Rebased all commits on top of `main`

**Comment by bdrhn9** [comment]
I found a chance to test it with our configuration. I can only say that it works a lot better than ever. Bunch of 🔥

**Comment by aquarat** [comment]
> I found a chance to test it with our configuration. I can only say that it works a lot better than ever. Bunch of 🔥

Thank you, that has made my day ❤️ 

I assume we're now waiting for @acenolaza to be happy. I'm excited about this 😊

**Comment by acenolaza** [comment]
> > I found a chance to test it with our configuration. I can only say that it works a lot better than ever. Bunch of fire
> 
> Thank you, that has made my day heart
> 
> I assume we're now waiting for @acenolaza to be happy. I'm excited about this blush

I'm running all my tests for the multicall PR using this branch as base. Everything looks good so far. Great work @aquarat  and thanks for testing this @bdrhn9 👏🏻

**Comment by aquarat** [comment]
We should probably try and merge this PR soon and then if there are any bugs fix them in new PRs 🙏 
We'll test this extensively when we run our testing beaconsets.

**Comment by aquarat** [comment]
Can this be merged now @acenolaza ?
I think we can do follow-up bug fixes in a new PR if needed 🤷

**Comment by acenolaza** [comment]
> Can this be merged now @acenolaza ? I think we can do follow-up bug fixes in a new PR if needed shrug

Yes, please go ahead 👍🏻

**Pull request merged**